File: _scripts/check_framework.mjs
/**
 * Content Framework Enforcer — runs all checks on modified blog posts.
 */
import { readFileSync, existsSync } from 'fs';

// ── Load data.js ────────────────────────────────────────────────────────────
const dataPath = new URL('../src/app/blog/data.js', import.meta.url).pathname;
const raw = readFileSync(dataPath, 'utf-8');

// Extract all post objects via a simple regex-based parse (the file is an array literal)
// We'll evaluate the JS directly using dynamic import of a temp file approach.
// Actually, let's use a safer eval approach.
import { writeFileSync, unlinkSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));

// Write a small wrapper that exports the array
const wrapperPath = join(__dirname, '___data_wrapper.mjs');
const wrapperContent = raw.replace('const posts = [', 'const posts = [\n');
writeFileSync(wrapperPath, wrapperContent + '\nexport { posts };\n');

let posts;
try {
  const mod = await import(/* fileURL */ `file://${wrapperPath}`);
  posts = mod.posts;
} catch (e) {
  console.error('Failed to parse data.js:', e.message);
  process.exit(1);
}

// Cleanup
try { unlinkSync(wrapperPath); } catch {}

// ── Modified slugs (from git log) ───────────────────────────────────────────
const modifiedSlugs = new Set([
  "seo-for-facebook-marketplace",
  "long-tail-keywords-bangladesh",
  "seo-tips-for-business-owners-bd",
  "seo-bangla-blog-content-writing",
  "seo-vs-google-ads-bangladesh-business",
  "youtube-seo-bangladesh-ranking-tips",
  "schema-markup-rich-snippets-techniques",
  "mobile-seo-bangladesh-ranking-strategy",
  "google-search-console-performance-guide",
  "content-marketing-seo-friendly-content-writing",
  "keyword-research-bangladesh-market",
  "link-building-bangladesh-strategies",
  "ecommerce-seo-daraz-shopify-guide",
  "technical-seo-core-web-vitals-optimization",
  "seo-trends-2026-ai-geo-future",
  "local-seo-dhaka-google-maps-ranking",
  "seo-bangla-beginners-guide-google-ranking",
  "international-seo-bangladesh-exporters-global-buyers",
  "content-marketing-strategy-bangladeshi-brands-seo",
  "mobile-seo-optimization-bangladesh-mobile-first-era",
  "seo-real-estate-developers-dhaka",
  "seo-vs-google-ads-whats-best-bangladesh-businesses",
  "google-business-profile-optimization-guide-bangladesh",
  "seo-garments-textile-industry-b2b-lead-generation",
  "geo-optimization-prepare-business-ai-search",
  "link-building-strategies-bangladesh-market",
  "how-to-choose-right-seo-agency-bangladesh",
  "technical-seo-checklist-bangladeshi-websites",
  "why-ecommerce-store-needs-seo-bangladesh",
  "local-seo-tips-dhaka-businesses-google-maps",
  "complete-seo-guide-bangladesh-businesses-2026",
  "seo-expert-vs-seo-agency-dhaka",
  "top-10-seo-mistakes-dhaka-businesses-make",
  "what-does-an-seo-expert-do",
  "ai-seo-2026-dhaka-experts",
  "seo-case-study-dhaka",
  "seo-roi-vs-paid-ads-dhaka",
  "how-to-choose-best-seo-expert-dhaka"
]);

// ── Helper functions ────────────────────────────────────────────────────────

function extractKeyword(title) {
  // Extract first meaningful noun phrase from title
  // Strip common prefixes
  let t = title.replace(/^(Complete|Ultimate|The |How to|Why|What|When|Where|Essential|Top \d+ )/i, '');
  // Take first 2-3 words
  const words = t.split(/[\s,—–-]+/).filter(w => w.length > 2).slice(0, 4);
  return words.join(' ').toLowerCase();
}

function countOccurrences(text, keyword) {
  if (!keyword || keyword.length < 3) return 999; // skip if not meaningful
  const words = keyword.toLowerCase().split(/\s+/);
  let count = 0;
  const lowerText = text.toLowerCase();
  // Count the full phrase
  const regex = new RegExp(keyword.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'gi');
  const matches = lowerText.match(regex);
  return matches ? matches.length : 0;
}

function countQuestionHeadings(content) {
  // Count headings (## or ###) that start with question words
  const qWords = ['How', 'What', 'Why', 'When', 'Where', 'Can', 'Do', 'Is', 'Are', 'Does', 'Which', 'Who'];
  const headingLines = content.match(/^#{2,3}\s+.*$/gm) || [];
  return headingLines.filter(h => {
    const cleanH = h.replace(/^#+\s*/, '');
    return qWords.some(q => cleanH.startsWith(q));
  }).length;
}

function countInternalLinks(content) {
  // Count markdown links starting with /
  const links = content.match(/\[([^\]]*)\]\((\/[^)]*)\)/g) || [];
  return links.length;
}

function findPillarLink(content, tags) {
  // Check if post links to a pillar page based on tags
  const tagStr = (tags || []).join(' ').toLowerCase();
  const contentStr = content.toLowerCase();
  
  // Map of tag-based pillar pages
  const pillarMap = [
    { tags: ['seo', 'seo guide', 'bangladesh seo', 'seo strategy'], pillar: '/blog/complete-seo-guide-bangladesh-businesses-2026', name: 'Complete SEO Guide' },
    { tags: ['local seo', 'google maps'], pillar: '/blog/local-seo-tips-dhaka-businesses-google-maps', name: 'Local SEO Dhaka guide' },
    { tags: ['ecommerce', 'daraz', 'shopify'], pillar: '/industries/ecommerce', name: 'E-commerce industry page' },
    { tags: ['technical seo', 'core web vitals', 'page speed'], pillar: '/blog/technical-seo-checklist-bangladeshi-websites', name: 'Technical SEO checklist' },
    { tags: ['link building'], pillar: '/blog/link-building-bangladesh-strategies', name: 'Link building guide' },
    { tags: ['keyword research', 'keywords'], pillar: '/blog/keyword-research-bangladesh-market', name: 'Keyword research guide' },
    { tags: ['content marketing'], pillar: '/industries/digital-marketing', name: 'Digital Marketing page' },
    { tags: ['geo', 'ai search', 'generative ai'], pillar: '/blog/geo-optimization-prepare-business-ai-search', name: 'GEO guide' },
    { tags: ['mobile seo', 'mobile'], pillar: '/blog/mobile-seo-optimization-bangladesh-mobile-first-era', name: 'Mobile SEO guide' },
    { tags: ['real estate', 'property'], pillar: '/industries/real-estate', name: 'Real estate industry page' },
    { tags: ['google ads', 'ppc', 'paid ads'], pillar: '/blog/seo-vs-google-ads-whats-best-bangladesh-businesses', name: 'SEO vs Google Ads guide' },
    { tags: ['google business profile', 'gbp'], pillar: '/blog/google-business-profile-optimization-guide-bangladesh', name: 'GBP guide' },
    { tags: ['garments', 'textile', 'b2b', 'manufacturing'], pillar: '/industries/manufacturing', name: 'Manufacturing page' },
    { tags: ['international seo', 'global', 'export'], pillar: '/blog/international-seo-bangladesh-exporters-global-buyers', name: 'International SEO guide' },
    { tags: ['beginners', 'beginner guide'], pillar: '/blog/seo-bangla-beginners-guide-google-ranking', name: 'SEO beginners guide' },
    { tags: ['schema', 'structured data', 'rich snippets'], pillar: '/blog/schema-markup-rich-snippets-techniques', name: 'Schema markup guide' },
    { tags: ['youtube', 'video seo'], pillar: '/blog/youtube-seo-bangladesh-ranking-tips', name: 'YouTube SEO guide' },
    { tags: ['google search console', 'gsc'], pillar: '/blog/google-search-console-performance-guide', name: 'Google Search Console guide' },
    { tags: ['seo agency', 'seo expert', 'agency'], pillar: '/blog/how-to-choose-right-seo-agency-bangladesh', name: 'Choose SEO agency guide' },
    { tags: ['facebook', 'social media', 'marketplace'], pillar: '/industries/ecommerce', name: 'E-commerce page' },
  ];
  
  for (const mapping of pillarMap) {
    const hasTag = mapping.tags.some(t => tagStr.includes(t));
    if (hasTag) {
      const linksToPillar = contentStr.includes(mapping.pillar.toLowerCase());
      return { pillarName: mapping.name, pillarUrl: mapping.pillar, found: linksToPillar };
    }
  }
  
  return { pillarName: 'Unknown', pillarUrl: '', found: false };
}

function checkEntities(content, title, tags) {
  const contentLower = content.toLowerCase();
  const missing = [];
  
  // Location check
  if (!contentLower.includes('dhaka') && !contentLower.includes('bangladesh')) {
    missing.push('Dhaka/Bangladesh location mention');
  }
  
  // Service type check - extract from tags
  const tagStr = (tags || []).join(' ').toLowerCase();
  if (tagStr.includes('seo') && !contentLower.includes('seo')) {
    missing.push('SEO service mention');
  }
  if (tagStr.includes('local') && !contentLower.includes('local')) {
    missing.push('Local SEO mention');
  }
  if (tagStr.includes('ecommerce') && !contentLower.includes('ecommerce') && !contentLower.includes('e-commerce')) {
    missing.push('E-commerce mention');
  }
  if (tagStr.includes('mobile') && !contentLower.includes('mobile')) {
    missing.push('Mobile SEO mention');
  }
  if (tagStr.includes('technical') && !contentLower.includes('technical')) {
    missing.push('Technical SEO mention');
  }
  if (tagStr.includes('link building') && !contentLower.includes('link building') && !contentLower.includes('link-build')) {
    missing.push('Link building mention');
  }
  if (tagStr.includes('keyword') && !contentLower.includes('keyword')) {
    missing.push('Keyword research mention');
  }
  
  return missing;
}

// ── Run checks ──────────────────────────────────────────────────────────────
const results = [];

for (const slug of modifiedSlugs) {
  const post = posts.find(p => p.slug === slug);
  if (!post) {
    results.push({ slug, error: 'Not found in data.js' });
    continue;
  }
  
  const { title, date, excerpt, tags = [], content = '', metaTitle, metaDescription, dateModified, imagePlaceholder } = post;
  const keyword = extractKeyword(title);
  const kwCount = countOccurrences(content, keyword);
  const questionHeadings = countQuestionHeadings(content);
  const internalLinks = countInternalLinks(content);
  const missingEntities = checkEntities(content, title, tags);
  const pillarInfo = findPillarLink(content, tags);
  
  // Schema check
  const schemaMissing = [];
  if (!title) schemaMissing.push('title');
  if (!excerpt) schemaMissing.push('excerpt');
  if (!date) schemaMissing.push('date');
  if (!metaTitle) schemaMissing.push('metaTitle');
  if (!metaDescription) schemaMissing.push('metaDescription');
  
  const tfIdfOk = kwCount >= 5 || keyword.length < 3;
  const entitiesOk = missingEntities.length === 0;
  const pillarOk = pillarInfo.found;
  const aeoOk = questionHeadings >= 2;
  const linksOk = internalLinks >= 3;
  const schemaOk = schemaMissing.length === 0;
  
  results.push({
    slug,
    title,
    keyword,
    kwCount,
    tfIdfOk,
    missingEntities,
    entitiesOk,
    pillarInfo,
    pillarOk,
    questionHeadings,
    aeoOk,
    internalLinks,
    linksOk,
    schemaMissing,
    schemaOk,
    hasContent: content.length > 50
  });
}

// ── Output report ────────────────────────────────────────────────────────────
console.log('# Content Framework Enforcement Report');
console.log(`Run: ${new Date().toISOString()}`);
console.log(`Posts checked: ${results.length}\n`);
console.log('---\n');

let allPassed = true;
let totalChecks = 0;
let passedChecks = 0;

for (const r of results) {
  if (r.error) {
    console.log(`## Post: ${r.slug}`);
    console.log('⚠️ **Not found in data.js**\n');
    continue;
  }
  
  const checks = [
    { name: `TF-IDF: "${r.keyword}"`, ok: r.tfIdfOk, detail: r.tfIdfOk ? `${r.kwCount} occurrences` : `Only ${r.kwCount} occurrences (need ≥5)` },
    { name: 'Entities', ok: r.entitiesOk, detail: r.entitiesOk ? 'All key entities present' : `Missing: ${r.missingEntities.join(', ')}` },
    { name: 'Pillar Link', ok: r.pillarOk, detail: r.pillarOk ? `Links to: ${r.pillarInfo.pillarName} (${r.pillarInfo.pillarUrl})` : `No link to pillar: ${r.pillarInfo.pillarName}` },
    { name: 'AEO/GEO', ok: r.aeoOk, detail: r.aeoOk ? `${r.questionHeadings} question headings` : `Only ${r.questionHeadings} question headings (need ≥2)` },
    { name: 'Internal Links', ok: r.linksOk, detail: r.linksOk ? `${r.internalLinks} internal links` : `Only ${r.internalLinks} internal links (need ≥3)` },
    { name: 'Schema Ready', ok: r.schemaOk, detail: r.schemaOk ? 'All fields set' : `Missing: ${r.schemaMissing.join(', ')}` },
  ];
  
  const hasFail = checks.some(c => !c.ok);
  if (hasFail) allPassed = false;
  
  for (const c of checks) {
    totalChecks++;
    if (c.ok) passedChecks++;
  }
  
  const statusIcon = hasFail ? '⚠️' : '✅';
  console.log(`## ${statusIcon} Post: ${r.slug}`);
  console.log(`**Title:** ${r.title}`);
  console.log('| Check | Status | Details |');
  console.log('|-------|--------|---------|');
  for (const c of checks) {
    console.log(`| ${c.name} | ${c.ok ? '✅' : '❌'} | ${c.detail} |`);
  }
  
  if (hasFail) {
    console.log('\n### Fix instructions:');
    if (!r.tfIdfOk) console.log(`- **TF-IDF**: Increase "${r.keyword}" occurrences to ≥5. Add a few more uses in body paragraphs.`);
    if (!r.entitiesOk) console.log(`- **Missing entities**: Add mentions of: ${r.missingEntities.join(', ')}`);
    if (!r.pillarOk) console.log(`- **Pillar link**: Add an internal link to ${r.pillarInfo.pillarName} (${r.pillarInfo.pillarUrl})`);
    if (!r.aeoOk) console.log(`- **AEO/GEO**: Add ≥${2 - r.questionHeadings} more question-based headings (How, What, Why, etc.)`);
    if (!r.linksOk) console.log(`- **Internal links**: Add at least ${3 - r.internalLinks} more internal links (to services, locations, or other posts)`);
    if (!r.schemaOk) console.log(`- **Schema**: Set missing fields: ${r.schemaMissing.join(', ')}`);
    if (!r.hasContent) console.log(`- **Content**: Post body appears empty or missing`);
    console.log('');
  } else {
    console.log('\n✅ All checks passed.\n');
  }
  console.log('---\n');
}

// ── Summary ─────────────────────────────────────────────────────────────────
console.log('## Overall Summary');
console.log(`| Metric | Value |`);
console.log(`|--------|-------|`);
console.log(`| Posts checked | ${results.length} |`);
console.log(`| Checks passed | ${passedChecks}/${totalChecks} (${Math.round(passedChecks/totalChecks*100)}%) |`);
console.log(`| All checks passed | ${allPassed ? '✅ Yes' : '❌ No'} |`);
const fails = results.filter(r => !r.error && (
  !r.tfIdfOk || !r.entitiesOk || !r.pillarOk || !r.aeoOk || !r.linksOk || !r.schemaOk
));
console.log(`| Posts needing fixes | ${fails.length} |`);
console.log('');
if (fails.length > 0) {
  console.log('### Posts needing attention:');
  for (const f of fails) {
    const issues = [];
    if (!f.tfIdfOk) issues.push('TF-IDF');
    if (!f.entitiesOk) issues.push('Entities');
    if (!f.pillarOk) issues.push('Pillar');
    if (!f.aeoOk) issues.push('AEO/GEO');
    if (!f.linksOk) issues.push('Internal Links');
    if (!f.schemaOk) issues.push('Schema');
    console.log(`- **${f.slug}** — ${issues.join(', ')}`);
  }
}

File: scripts/check_framework.py
#!/usr/bin/env python3
"""
Content Framework Enforcer for kanokmiah.com.bd
Checks each blog post against the 6-point content framework.
"""

import re
import json
import sys

# Read the data.js file
with open('/root/kanok-miahit/src/app/blog/data.js', 'r') as f:
    content = f.read()

# Parse out each post using regex
# Posts are objects in the posts array
post_pattern = re.compile(
    r'\{[^}]*slug:\s*"([^"]+)"[^}]*title:\s*"([^"]*)"[^}]*date:\s*"([^"]*)"[^}]*excerpt:\s*"([^"]*)"[^}]*tags:\s*\[([^\]]*)\][^}]*content:\s*`([^`]*)`',
    re.DOTALL
)

# Alternative: split by slug and extract manually
# Let me use a simpler approach
posts_raw = re.findall(r'\{[^{}]*slug:\s*"([^"]+)"[^{}]*\}', content)
print(f"Found {len(posts_raw)} posts via regex", file=sys.stderr)

# Better approach: parse slug by slug
slug_positions = []
for m in re.finditer(r'slug:\s*"([^"]+)"', content):
    slug_positions.append((m.start(), m.group(1)))

print(f"Found {len(slug_positions)} slugs", file=sys.stderr)

# Get post boundaries: each post starts with `{` and ends with `},`
# Let's extract each post section
post_sections = []
for i, (pos, slug) in enumerate(slug_positions):
    # Find the start of this post object
    start = content.rfind('{', 0, pos)
    if start == -1:
        continue
    
    # Find the end of this post object (the matching `},` or `]` for the last one)
    if i < len(slug_positions) - 1:
        next_pos = slug_positions[i+1][0]
        # Find the `},` before the next slug
        end_match = content.rfind('},', pos, next_pos)
        if end_match == -1:
            end_match = content.rfind('`,\n  {', pos, next_pos)
        if end_match == -1:
            end = next_pos
        else:
            end = end_match + 2
    else:
        # Last post - find the closing
        end = content.find('\n];', pos)
        if end == -1:
            end = len(content)
    
    post_text = content[start:end]
    post_sections.append((slug, start, end, post_text))

# For each post, extract the full data including content inside backticks
# Let me use a more robust approach
posts = []
# Find all post objects
pattern = re.compile(
    r'\{[\s\S]*?slug:\s*"([^"]+)"[\s\S]*?title:\s*"([^"]*)"[\s\S]*?date:\s*"([^"]*)"[\s\S]*?excerpt:\s*"([^"]*)"[\s\S]*?tags:\s*\[([^\]]*)\][\s\S]*?content:\s*`\n?([\s\S]*?)`\s*,',
    re.DOTALL
)

matches = list(pattern.finditer(content))
print(f"Parsed {len(matches)} posts with full content", file=sys.stderr)

results = []

for m in matches[:]:  # Use all posts
    slug = m.group(1)
    title = m.group(2)
    date = m.group(3)
    excerpt = m.group(4)
    tags_str = m.group(5)
    post_content = m.group(6)
    
    tags = [t.strip().strip('"') for t in tags_str.split(',') if t.strip()]
    
    # ============ A. TF-IDF Coverage ============
    # Extract primary keyword from title (first meaningful noun phrase)
    title_lower = title.lower()
    # Remove "Complete", "Guide", "for", "in", "the", "a", "an" etc. at start to find the main topic
    stop_words = {'complete', 'guide', 'for', 'the', 'a', 'an', 'in', 'to', 'of', 'is', 'what', 'how', 'why', 'when', 'where', 'best', 'top', 'ultimate', 'essential', 'comprehensive', 'your', 'our'}
    
    # Extract meaningful words from title
    title_words = [w for w in title_lower.split() if w not in stop_words and len(w) > 2]
    
    # Pick primary keyword: first 2-3 meaningful words
    if len(title_words) >= 3:
        primary_kw = ' '.join(title_words[:3])
    elif len(title_words) >= 2:
        primary_kw = ' '.join(title_words[:2])
    elif len(title_words) >= 1:
        primary_kw = title_words[0]
    else:
        primary_kw = title_lower
    
    # Special case: if title starts with a question word, use the noun phrase after it
    question_starts = ['what is', 'what are', 'how to', 'how do', 'why is', 'why does']
    for qs in question_starts:
        if title_lower.startswith(qs):
            remainder = title_lower[len(qs):].strip()
            words = [w for w in remainder.split() if w not in stop_words and len(w) > 2]
            if words:
                primary_kw = ' '.join(words[:3]) if len(words) >= 3 else ' '.join(words[:2])
            break
    
    # Also check for SEO-related keywords - use the first prominent topic
    # Count occurrences
    kw_lower = primary_kw.lower()
    count = post_content.lower().count(kw_lower)
    
    tfidf_status = "✅" if count >= 5 else "❌"
    tfidf_detail = f"'{primary_kw}' → {count} occurrences"
    
    # ============ B. Semantic Entity Coverage ============
    required_entities = [
        ("location_dhaka", "Dhaka"),
        ("location_bangladesh", "Bangladesh"),
        ("service_seo", "SEO"),
    ]
    
    # Check tags for industry hints
    has_ecommerce = any('ecommerce' in t.lower() or 'shopify' in t.lower() or 'daraz' in t.lower() for t in tags)
    has_local = any('local' in t.lower() for t in tags)
    has_technical = any('technical' in t.lower() for t in tags)
    has_realestate = any('real estate' in t.lower() for t in tags)
    has_healthcare = any('healthcare' in t.lower() or 'medical' in t.lower() or 'clinic' in t.lower() for t in tags)
    has_education = any('education' in t.lower() for t in tags)
    has_hotel = any('hotel' in t.lower() or 'resort' in t.lower() or 'travel' in t.lower() or 'tourism' in t.lower() for t in tags)
    has_garment = any('garment' in t.lower() or 'textile' in t.lower() for t in tags)
    has_bangla = any('bangla' in t.lower() or 'বাংলা' in t.lower() for t in tags)
    has_linkbuilding = any('link building' in t.lower() or 'backlink' in t.lower() for t in tags)
    has_mobile = any('mobile' in t.lower() for t in tags)
    has_content = any('content' in t.lower() for t in tags)
    has_restaurant = any('restaurant' in t.lower() or 'cafe' in t.lower() or 'food' in t.lower() for t in tags)
    has_cleaning = any('cleaning' in t.lower() for t in tags)
    has_ngo = any('ngo' in t.lower() or 'non-profit' in t.lower() or 'nonprofit' in t.lower() for t in tags)
    has_law = any('law' in t.lower() or 'legal' in t.lower() or 'firm' in t.lower() for t in tags)
    has_fitness = any('fitness' in t.lower() or 'gym' in t.lower() for t in tags)
    has_startup = any('startup' in t.lower() for t in tags)
    has_photography = any('photographer' in t.lower() or 'videographer' in t.lower() for t in tags)
    has_wedding = any('wedding' in t.lower() or 'event' in t.lower() for t in tags)
    has_b2b = any('b2b' in t.lower() or 'lead generation' in t.lower() for t in tags)
    has_schema = any('schema' in t.lower() or 'structured data' in t.lower() for t in tags)
    has_voice = any('voice' in t.lower() for t in tags)
    has_case_study = 'case-study' in slug
    
    # Check entities present
    content_lower = post_content.lower()
    
    missing_entities = []
    
    if not has_bangla:
        if 'dhaka' not in content_lower:
            missing_entities.append("Dhaka")
        if 'bangladesh' not in content_lower:
            missing_entities.append("Bangladesh")
    
    if 'seo' not in content_lower:
        missing_entities.append("SEO (service)")
    
    # Industry-specific entities
    if has_ecommerce and 'e-commerce' not in content_lower and 'ecommerce' not in content_lower and 'online store' not in content_lower:
        missing_entities.append("e-commerce (industry)")
    if has_local and ('local seo' not in content_lower) and ('google maps' not in content_lower) and ('google business profile' not in content_lower):
        missing_entities.append("local SEO / GBP")
    if has_technical and ('technical seo' not in content_lower) and ('core web vitals' not in content_lower):
        missing_entities.append("technical SEO / Core Web Vitals")
    if has_garment and ('garment' not in content_lower and 'textile' not in content_lower):
        missing_entities.append("garment/textile (industry)")
    if has_restaurant and ('restaurant' not in content_lower and 'cafe' not in content_lower and 'food' not in content_lower):
        missing_entities.append("restaurant/food (industry)")
    if has_healthcare and ('healthcare' not in content_lower and 'medical' not in content_lower and 'clinic' not in content_lower and 'doctor' not in content_lower):
        missing_entities.append("healthcare/medical (industry)")
    if has_education and ('education' not in content_lower and 'school' not in content_lower and 'university' not in content_lower and 'student' not in content_lower and 'educational' not in content_lower):
        missing_entities.append("education (industry)")
    if has_realestate and ('real estate' not in content_lower and 'property' not in content_lower and 'developer' not in content_lower):
        missing_entities.append("real estate (industry)")
    if has_hotel and ('hotel' not in content_lower and 'resort' not in content_lower and 'travel' not in content_lower and 'tourism' not in content_lower):
        missing_entities.append("hotel/travel (industry)")
    if has_linkbuilding and ('link building' not in content_lower and 'backlink' not in content_lower):
        missing_entities.append("link building/backlinks")
    if has_mobile and ('mobile seo' not in content_lower and 'mobile optimization' not in content_lower):
        missing_entities.append("mobile SEO")
    if has_content and ('content marketing' not in content_lower and 'content strategy' not in content_lower):
        missing_entities.append("content marketing (method)")
    if has_cleaning and ('cleaning' not in content_lower):
        missing_entities.append("cleaning (industry)")
    if has_ngo and ('ngo' not in content_lower and 'non-profit' not in content_lower and 'nonprofit' not in content_lower and 'charity' not in content_lower and 'social' not in content_lower):
        missing_entities.append("NGO/non-profit (industry)")
    if has_law and ('law' not in content_lower and 'legal' not in content_lower and 'attorney' not in content_lower and 'lawyer' not in content_lower and 'firm' not in content_lower):
        missing_entities.append("legal/law (industry)")
    if has_fitness and ('fitness' not in content_lower and 'gym' not in content_lower and 'workout' not in content_lower):
        missing_entities.append("fitness/gym (industry)")
    if has_startup and ('startup' not in content_lower):
        missing_entities.append("startup (industry)")
    if has_photography and ('photographer' not in content_lower and 'photography' not in content_lower and 'videographer' not in content_lower):
        missing_entities.append("photography (industry)")
    if has_wedding and ('wedding' not in content_lower and 'event' not in content_lower and 'planner' not in content_lower):
        missing_entities.append("wedding/event (industry)")
    if has_b2b and ('b2b' not in content_lower and 'lead generation' not in content_lower):
        missing_entities.append("B2B/lead generation")
    if has_schema and ('schema' not in content_lower and 'structured data' not in content_lower and 'rich snippet' not in content_lower):
        missing_entities.append("schema/structured data")
    if has_voice and ('voice search' not in content_lower and 'voice' not in content_lower):
        missing_entities.append("voice search")
    
    entities_status = "✅" if len(missing_entities) == 0 else "❌"
    entities_detail = f"Missing: {', '.join(missing_entities)}" if missing_entities else "All key entities present"
    
    # ============ C. Pillar-Cluster Alignment ============
    pillar_pages = [
        "/blog/complete-seo-guide-bangladesh-businesses-2026",
        "/services/seo",
        "/services/local-seo",
        "/services/technical-seo",
        "/services/seo-consulting",
    ]
    
    has_pillar_link = any(f'({p}' in post_content or f'href="{p}' in post_content for p in pillar_pages)
    
    # Check if it links to its pillar based on tags
    pillar_link_status = "✅" if has_pillar_link else "❌"
    pillar_link_detail = "No pillar link found" if not has_pillar_link else "Links to pillar page"
    
    # ============ D. AEO/GEO Optimization ============
    question_heads = re.findall(r'^#{1,3}\s+(How\s|What\s|Why\s|When\s|Where\s|Can\s|Do\s|Is\s|Are\s|Does\s|Which\s|Who\s)', post_content, re.MULTILINE)
    num_question_heads = len(question_heads)
    
    aeo_status = "✅" if num_question_heads >= 2 else "❌"
    aeo_detail = f"{num_question_heads} question headings"
    
    # ============ E. Internal Linking ============
    # Count internal links (links starting with /)
    internal_links = re.findall(r'\(/(?:blog/|services/|locations/|industries/|about|contact|/)[^)]*\)', post_content)
    # Filter for actual links to other posts/pages (not self-references)
    self_link = f'/blog/{slug}'
    meaningful_links = [l for l in internal_links if self_link not in l]
    
    internal_links_status = "✅" if len(meaningful_links) >= 3 else "❌"
    internal_links_detail = f"{len(meaningful_links)} internal links"
    
    # ============ F. Schema Readiness ============
    schema_missing = []
    if not title:
        schema_missing.append("title")
    if not excerpt:
        schema_missing.append("excerpt")
    if not date:
        schema_missing.append("date")
    
    schema_status = "✅" if len(schema_missing) == 0 else "❌"
    schema_detail = f"Missing: {', '.join(schema_missing)}" if schema_missing else "All fields set"
    
    # ============ Assemble Result ============
    results.append({
        'slug': slug,
        'title': title,
        'checks': {
            'TF-IDF': {'status': tfidf_status, 'detail': tfidf_detail},
            'Entities': {'status': entities_status, 'detail': entities_detail},
            'Pillar Link': {'status': pillar_link_status, 'detail': pillar_link_detail},
            'AEO/GEO': {'status': aeo_status, 'detail': aeo_detail},
            'Internal Links': {'status': internal_links_status, 'detail': internal_links_detail},
            'Schema Ready': {'status': schema_status, 'detail': schema_detail},
        }
    })

# Output as JSON
print(json.dumps(results, indent=2))
